programs/rodeos/streams/rabbitmq.hpp

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include "stream.hpp"
+#include "amqpcpp.h"
+#include "amqpcpp/libboostasio.h"
+#include "amqpcpp/linux_tcp.h"
+#include <fc/log/logger.hpp>
+#include <memory>
+
+namespace b1 {
+
+class rabbitmq_handler;
+
+class rabbitmq : public stream_handler {
+   std::shared_ptr<rabbitmq_handler>    handler_;
+   std::shared_ptr<AMQP::TcpConnection> connection_;
+   std::shared_ptr<AMQP::TcpChannel>    channel_;
+   std::string                          name_;
+   std::vector<eosio::name>             routes_;
+
+ public:
+   rabbitmq(boost::asio::io_service& io_service, std::vector<eosio::name> routes, std::string address, std::string name)
+       : name_(name), routes_(routes) {
+      AMQP::Address amqp_address(address);
+      ilog("Connecting to RabbitMQ address ${a} - Queue: ${q}...", ("a", std::string(amqp_address))("q", name));
+
+      handler_    = std::make_shared<rabbitmq_handler>(io_service);
+      connection_ = std::make_shared<AMQP::TcpConnection>(handler_.get(), amqp_address);
+      channel_    = std::make_shared<AMQP::TcpChannel>(connection_.get());
+      declare_queue();
+   }
+
+   std::vector<eosio::name>& get_routes() { return routes_; }
+
+   void publish(const char* data, uint64_t data_size) { channel_->publish("", name_, data, data_size, 0); }
+
+ private:
+   void declare_queue() {
+      auto& queue = channel_->declareQueue(name_, AMQP::durable);
+      queue.onSuccess([](const std::string& name, uint32_t messagecount, uint32_t consumercount) {
+         ilog("RabbitMQ Connected Successfully!\n Queue ${q} - Messages: ${mc} - Consumers: ${cc}",
+              ("q", name)("mc", messagecount)("cc", consumercount));
+      });
+      queue.onError([](const char* error_message) {
+         throw std::runtime_error("RabbitMQ Queue error: " + std::string(error_message));
```

Comment:
If I'm looking at amqp-cpp correctly you should not throw from inside `onError`. I don't think amqp-cpp is setup to handle that correctly. Instead just `elog` the error.

Reply:
Apparently needed to shutdown amqp-cpp